libraries/AP_Baro/AP_Baro_BMP280.cpp

```diff
@@ -0,0 +1,172 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "AP_Baro_BMP280.h"
+
+#include <utility>
+
+#include <AP_HAL/AP_HAL.h>
+
+extern const AP_HAL::HAL &hal;
+
+#define BMP280_MODE_SLEEP  0
+#define BMP280_MODE_FORCED 1
+#define BMP280_MODE_NORMAL 3
+#define BMP280_MODE BMP280_MODE_NORMAL
+
+#define BMP280_OVERSAMPLING_1  1
+#define BMP280_OVERSAMPLING_2  2
+#define BMP280_OVERSAMPLING_4  3
+#define BMP280_OVERSAMPLING_8  4
+#define BMP280_OVERSAMPLING_16 5
+#define BMP280_OVERSAMPLING_P BMP280_OVERSAMPLING_16
+#define BMP280_OVERSAMPLING_T BMP280_OVERSAMPLING_2
+
+#define BMP280_FILTER_COEFFICIENT 2
+
+#define BMP280_ID            0x58
+
+#define BMP280_REG_ID        0xD0
+#define BMP280_REG_RESET     0xE0
+#define BMP280_REG_STATUS    0xF3
+#define BMP280_REG_CTRL_MEAS 0xF4
+#define BMP280_REG_CONFIG    0xF5
+#define BMP280_RED_DATA      0xF7
+
+AP_Baro_BMP280::AP_Baro_BMP280(AP_Baro &baro, AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+    : AP_Baro_Backend(baro)
+    , _dev(std::move(dev))
+{
+    uint8_t buf[24];
+
+    // Get pointer to I2C bus semaphore
+    AP_HAL::Semaphore *sem = _dev->get_semaphore();
+
+    // Take I2C bus sempahore
+    if (!sem->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        AP_HAL::panic("BMP280: unable to get semaphore");
+    }
+
+    // Check BMP280 ID
+    uint8_t val = 0;
+    
+    if (!_dev->read_registers(BMP280_REG_ID, &val, 1)) {
+        AP_HAL::panic("BMP280: can't read ID");
+    }
+    if (val != BMP280_ID) {
+        AP_HAL::panic("BMP280: wrong ID");
```

Comment:
for new drivers I would like to use the probe() and ADD_BACKEND() approach that we now use for the ms5611. That allows the sensor to be optionally detected on an external i2c bus (for example). It means not doing any of the probing in the constructor, and instead having a static probe() method. Otherwise it will panic if the sensor is not found.